### ISSUE TITLE
[crash] heap-use-after-free occurs when using -b and -E

### ISSUE THREAD

**Issue by Ruanxingzhi** [Orc/discount#281]
When using the program with -b and -E, a heap-use-after-free occurs.

This bug affects versions **2.x, 3.0.0a and 3.0.0b**.

### Reproduce

Run:
```bash
./markdown -b hello -E world -s "[a](b) [a](b)"   # this input breaks 2.x, 3.0.0a and 3.0.0b

./markdown -b hello -E world -s "[a](b)"   # this input breaks 2.x and 3.0.0a
```

Result:
```
=================================================================
==1791066==ERROR: AddressSanitizer: heap-use-after-free on address 0x602000000010 at pc 0x555555676df9 bp 0x7fffffffc9b0 sp 0x7fffffffc9a8
READ of size 1 at 0x602000000010 thread T0
    #0 0x555555676df8 in Qstring /work/programs/discount/src-1213-origin/generate.c:153:5
    #1 0x55555567ab23 in printlinkyref /work/programs/discount/src-1213-origin/generate.c:633:2
    #2 0x55555567c9b2 in linkyformat /work/programs/discount/src-1213-origin/generate.c:726:2
    #3 0x555555677224 in linkylinky /work/programs/discount/src-1213-origin/generate.c:773:12
    #4 0x555555673392 in text /work/programs/discount/src-1213-origin/generate.c:1391:14
    #5 0x5555556825da in printblock /work/programs/discount/src-1213-origin/generate.c:1810:5
    #6 0x55555567f9d3 in display /work/programs/discount/src-1213-origin/generate.c:2018:2
    #7 0x55555567f615 in htmlify_paragraphs /work/programs/discount/src-1213-origin/generate.c:1872:5
    #8 0x5555556755f5 in htmlify /work/programs/discount/src-1213-origin/generate.c:1916:5
    #9 0x55555567f9c1 in display /work/programs/discount/src-1213-origin/generate.c:2014:2
    #10 0x55555567f615 in htmlify_paragraphs /work/programs/discount/src-1213-origin/generate.c:1872:5
    #11 0x5555556755f5 in htmlify /work/programs/discount/src-1213-origin/generate.c:1916:5
    #12 0x555555674b3a in mkd_document /work/programs/discount/src-1213-origin/generate.c:2065:6
    #13 0x55555565bdbd in mkd_generatehtml /work/programs/discount/src-1213-origin/mkdio.c:200:5
    #14 0x555555654973 in main /work/programs/discount/src-1213-origin/main.c:381:7
    #15 0x7ffff7cea6c9  (/lib/x86_64-linux-gnu/libc.so.6+0x276c9) (BuildId: 8a1bf172e710f8ca0c1576912c057b45f90d90d8)
    #16 0x7ffff7cea784 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27784) (BuildId: 8a1bf172e710f8ca0c1576912c057b45f90d90d8)
    #17 0x55555557c3f0 in _start (/work/programs/discount/src-1213-origin/markdown+0x283f0) (BuildId: 6e6aac165e444e2f95f12e7d1550fcd9650848f5)

0x602000000010 is located 0 bytes inside of 6-byte region [0x602000000010,0x602000000016)
freed by thread T0 here:
    #0 0x555555616d9a in free (/work/programs/discount/src-1213-origin/markdown+0xc2d9a) (BuildId: 6e6aac165e444e2f95f12e7d1550fcd9650848f5)
    #1 0x55555568bf73 in e_free /work/programs/discount/src-1213-origin/basename.c:34:19
    #2 0x55555567ac4a in printlinkyref /work/programs/discount/src-1213-origin/generate.c:634:23
    #3 0x55555567c9b2 in linkyformat /work/programs/discount/src-1213-origin/generate.c:726:2
    #4 0x555555677224 in linkylinky /work/programs/discount/src-1213-origin/generate.c:773:12
    #5 0x555555673392 in text /work/programs/discount/src-1213-origin/generate.c:1391:14
    #6 0x5555556825da in printblock /work/programs/discount/src-1213-origin/generate.c:1810:5
    #7 0x55555567f9d3 in display /work/programs/discount/src-1213-origin/generate.c:2018:2
    #8 0x55555567f615 in htmlify_paragraphs /work/programs/discount/src-1213-origin/generate.c:1872:5
    #9 0x5555556755f5 in htmlify /work/programs/discount/src-1213-origin/generate.c:1916:5
    #10 0x55555567f9c1 in display /work/programs/discount/src-1213-origin/generate.c:2014:2
    #11 0x55555567f615 in htmlify_paragraphs /work/programs/discount/src-1213-origin/generate.c:1872:5
    #12 0x5555556755f5 in htmlify /work/programs/discount/src-1213-origin/generate.c:1916:5
    #13 0x555555674b3a in mkd_document /work/programs/discount/src-1213-origin/generate.c:2065:6
    #14 0x55555565bdbd in mkd_generatehtml /work/programs/discount/src-1213-origin/mkdio.c:200:5
    #15 0x555555654973 in main /work/programs/discount/src-1213-origin/main.c:381:7
    #16 0x7ffff7cea6c9  (/lib/x86_64-linux-gnu/libc.so.6+0x276c9) (BuildId: 8a1bf172e710f8ca0c1576912c057b45f90d90d8)

previously allocated by thread T0 here:
    #0 0x555555600e21 in strdup (/work/programs/discount/src-1213-origin/markdown+0xace21) (BuildId: 6e6aac165e444e2f95f12e7d1550fcd9650848f5)
    #1 0x5555556546ab in main /work/programs/discount/src-1213-origin/main.c:350:26
    #2 0x7ffff7cea6c9  (/lib/x86_64-linux-gnu/libc.so.6+0x276c9) (BuildId: 8a1bf172e710f8ca0c1576912c057b45f90d90d8)

SUMMARY: AddressSanitizer: heap-use-after-free /work/programs/discount/src-1213-origin/generate.c:153:5 in Qstring
Shadow bytes around the buggy address:
  0x601ffffffd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x602000000000: fa fa[fd]fa fa fa fd fa fa fa fa fa fa fa fa fa
  0x602000000080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1791066==ABORTING
```

**Comment by Orc** [comment]
Holy christ.   I looked at my callback code and oh god it's a hot mess of a disaster.

I don't know what I was thinking when I wrote it.    Thank you for pointing out the biohazard area!

**Comment by Orc** [comment]
... and a little later I've redone the internals of the callbacks, and it doesn't seem to be tripping over anything obvious when I run the batch of testcases you've reported to me.  3.0.0c  is the new version#.

**Comment by Ruanxingzhi** [comment]
This issue is fixed in deca3b64b074192e33c21ae2c03e90437c604024. How about #280?
